crates/telio-starcast/src/multicaster.rs

```diff
@@ -0,0 +1,281 @@
+                format!("Failed to bind multicast socket: {:?}", e)
+            })?;
+        let multicast_socket = Arc::new(multicast_socket);
+
+        Ok(Self {
+            wg,
+            multicast_socket,
+            packet_chan: Some(packet_chan),
+            vpeer_ip,
+            listener_task: None,
+        })
+    }
+
+    pub async fn start_multicaster(&mut self) {
+        let Some(packet_chan) = self.packet_chan.take() else {
+            telio_log_warn!("multicast can be started once for now");
+            return;
+        };
+
+        self.listener_task = Some(Task::start(MulticastListener::new(
+            self.wg.clone(),
+            self.vpeer_ip,
+            self.multicast_socket.clone(),
+            packet_chan,
+        )));
+    }
+
+    pub async fn stop(&mut self) {
+        if let Some(listener_task) = self.listener_task.take() {
+            let _ = listener_task.stop().await.resume_unwind();
+        }
+    }
+}
+
+enum PacketType {
+    LocalMulticast,
+    MappedUnicast,
+    Misc,
+    Unsupported,
+}
+
+enum PacketOrigin {
+    Socket,
+    Channel,
+}
+
+struct MulticastListener {
+    wg: Arc<Interface>,
+    vpeer_ip: Ipv4Addr,
+    multicast_socket: Arc<UdpSocket>,
+    packet_chan: Chan<Vec<u8>>,
+    recv_buffer: Vec<u8>,
+    nat_map: RwLock<HashMap<u16, SocketAddr>>,
+}
+
+impl MulticastListener {
+    fn new(
+        wg: Arc<Interface>,
+        vpeer_ip: Ipv4Addr,
+        multicast_socket: Arc<UdpSocket>,
+        packet_chan: Chan<Vec<u8>>,
+    ) -> Self {
+        Self {
+            wg,
+            vpeer_ip,
+            multicast_socket,
+            packet_chan,
+            recv_buffer: vec![0; MAX_PACKET_SIZE],
+            nat_map: RwLock::new(HashMap::new()),
+        }
+    }
+
+    async fn route_packet(&mut self, mut packet: Vec<u8>, origin: PacketOrigin) {
+        let packet_type = if let Some(ip_packet) = MutableIpv4Packet::new(&mut packet) {
+            let packet_type =
+                self.identify_packet_type(IpAddr::V4(ip_packet.get_destination()), origin);
+            Some(packet_type)
+        } else {
+            None
+        };
+        match packet_type {
+            Some(PacketType::LocalMulticast) => self.handle_local_multicast_packet(packet).await,
+            Some(PacketType::MappedUnicast) => self.handle_mapped_unicast_packet(packet).await,
+            Some(PacketType::Misc) => self.handle_misc_packet(packet).await,
+            Some(PacketType::Unsupported) | None => {}
+        }
+    }
+
+    fn identify_packet_type(&self, dst_addr: IpAddr, origin: PacketOrigin) -> PacketType {
+        match (origin, dst_addr) {
+            (_, IpAddr::V6(_)) => {
+                telio_log_debug!("MulticastListener: IPv6 is not yet supported");
+                PacketType::Unsupported
+            }
+            (PacketOrigin::Channel, _) => PacketType::MappedUnicast,
+            (PacketOrigin::Socket, IpAddr::V4(addr)) if addr.octets()[0] == 244 => {
```

Comment:
So multicast is 224.0.0.0/4, so this check is highly incaccurate, also I will need to hash this out a bit, but we will probably just use a subset of that also. So it would be could to make this an argument to entire compomentent, to make it easier i recommned using: Vec<https://docs.rs/ipnetwork/latest/ipnetwork/enum.IpNetwork.html>, we also will need to pass this to vpeer to change it's config.

Reply:
`244` was a typo, it was supposed to be `224`, but using `IpNetwork` is a better option either way